### ISSUE TITLE
EC2 Instance?

### ISSUE THREAD

**Issue by davidcranor** [jcjohnson/neural-style#48]
Hello,

Is there an EC2 instance with this stuff loaded on it?  I'm new at this and have gotten it to compile and run on my GPU-less computer, but I crave more speed.  I have tried instance ami-ffba7b94 from here:

http://blog.titocosta.com/post/110345699197/public-ec2-ami-with-torch-and-caffe-deep-learning

But it neural_style is unhappy.  Appended below is the output when I tried to run the test command in GPU mode after installing neural_style.

If there is no EC2 instance with neural_style on it, any ideas of what I should do to fix this run error?

Thanks!
David

*****************

'ubuntu@ip-172-31-2-130:~/neural-style$ th neural_style.lua -gpu 0 -print_iter 1
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:505] Reading dangerously large protocol message.  If the message turns out to be larger than 1073741824 bytes, parsing will be halted for security reasons.  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:78] The total number of bytes read was 574671192
Successfully loaded models/VGG_ILSVRC_19_layers.caffemodel
/home/ubuntu/torch-distro/install/bin/luajit: ./loadcaffe_wrapper.lua:57: bad argument #1 to 'ipairs' (table expected, got nil)
stack traceback:
	[C]: in function 'ipairs'
	./loadcaffe_wrapper.lua:57: in function 'load'
	neural_style.lua:59: in function 'main'
	neural_style.lua:437: in main chunk
	[C]: in function 'dofile'
	...rch-distro/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
	[C]: at 0x00406670'

**Comment by hughperkins** [comment]
some things you can try, from a `th` session:
```
require 'cutorch'
require 'cunn'
a = torch.CudaTensor(3,2):uniform
print(a+1)
nn.SpatialConvolutionMM(3,3,1,1):cuda()
nn.Relu():cuda()
nn.SpatialMaxPooling(3,3,2,2):cuda()
```
Might throw some light on whats missing maybe?

**Comment by davidcranor** [comment]
hmm, all of those commands worked.

i think that something else is broken - even when I try to run it in CPU mode it still breaks.  Is it something with caffe?  This instance ostensibly has it....but I am unfamiliar with the interdependencies of all of these tools so am not sure what to check/recompile :-(

***********

ubuntu@ip-172-31-9-105:~/neural-style$ th neural_style.lua -gpu -1 -print_iter -1
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:505] Reading dangerously large protocol message.  If the message turns out to be larger than 1073741824 bytes, parsing will be halted for security reasons.  To increase the limit (or to disable these warnings), see CodedInputStream::SetTotalBytesLimit() in google/protobuf/io/coded_stream.h.
[libprotobuf WARNING google/protobuf/io/coded_stream.cc:78] The total number of bytes read was 574671192
Successfully loaded models/VGG_ILSVRC_19_layers.caffemodel
/home/ubuntu/torch-distro/install/bin/luajit: models/VGG_ILSVRC_19_layers_deploy.prototxt.cpu.lua:7: attempt to index global 'inn' (a nil value)
stack traceback:
	models/VGG_ILSVRC_19_layers_deploy.prototxt.cpu.lua:7: in main chunk
	[C]: in function 'dofile'
	./loadcaffe_wrapper.lua:48: in function 'load'
	neural_style.lua:59: in function 'main'
	neural_style.lua:437: in main chunk
	[C]: in function 'dofile'
	...rch-distro/install/lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
	[C]: at 0x00406670

**Comment by jcjohnson** [comment]
neural-style doesn't actually require caffe to be installed, but you do need to have the protobuf compiler installed.

Looks like the problems is that the file `models/VGG_ILSVRC_19_layers_deploy.prototxt.cpu.lua` (which contains the torch structure of the VGG-19 network) is trying to import the `inn` torch package which is not installed. You can install it by running `luarocks install inn`.

However this is confusing; the dependency on inn when running in CPU mode should have been fixed way back in https://github.com/jcjohnson/neural-style/commit/cba886c6c33ec53e4a6a56a67d5edf304cee88a0. Are you using a really old version of neural-style?

**Comment by davidcranor** [comment]
Hmm, i installed inn and it still errors out.  I'm using a freshly cloned neural-style every time I spin up EC2 to work with this stuff.

**Comment by jcjohnson** [comment]
Weird. Can you post your `models/VGG_ILSVRC_19_layers_deploy.prototxt.cpu.lua` file?

**Comment by davidcranor** [comment]
equire 'nn'
model = {}
table.insert(model, {'conv1_1', nn.SpatialConvolutionMM(3, 64, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu1_1', nn.ReLU()})
table.insert(model, {'conv1_2', nn.SpatialConvolutionMM(64, 64, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu1_2', nn.ReLU()})
table.insert(model, {'pool1', inn.SpatialMaxPooling(2, 2, 2, 2)})
table.insert(model, {'conv2_1', nn.SpatialConvolutionMM(64, 128, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu2_1', nn.ReLU()})
table.insert(model, {'conv2_2', nn.SpatialConvolutionMM(128, 128, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu2_2', nn.ReLU()})
table.insert(model, {'pool2', inn.SpatialMaxPooling(2, 2, 2, 2)})
table.insert(model, {'conv3_1', nn.SpatialConvolutionMM(128, 256, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu3_1', nn.ReLU()})
table.insert(model, {'conv3_2', nn.SpatialConvolutionMM(256, 256, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu3_2', nn.ReLU()})
table.insert(model, {'conv3_3', nn.SpatialConvolutionMM(256, 256, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu3_3', nn.ReLU()})
table.insert(model, {'conv3_4', nn.SpatialConvolutionMM(256, 256, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu3_4', nn.ReLU()})
table.insert(model, {'pool3', inn.SpatialMaxPooling(2, 2, 2, 2)})
table.insert(model, {'conv4_1', nn.SpatialConvolutionMM(256, 512, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu4_1', nn.ReLU()})
table.insert(model, {'conv4_2', nn.SpatialConvolutionMM(512, 512, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu4_2', nn.ReLU()})
table.insert(model, {'conv4_3', nn.SpatialConvolutionMM(512, 512, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu4_3', nn.ReLU()})
table.insert(model, {'conv4_4', nn.SpatialConvolutionMM(512, 512, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu4_4', nn.ReLU()})
table.insert(model, {'pool4', inn.SpatialMaxPooling(2, 2, 2, 2)})
table.insert(model, {'conv5_1', nn.SpatialConvolutionMM(512, 512, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu5_1', nn.ReLU()})
table.insert(model, {'conv5_2', nn.SpatialConvolutionMM(512, 512, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu5_2', nn.ReLU()})
table.insert(model, {'conv5_3', nn.SpatialConvolutionMM(512, 512, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu5_3', nn.ReLU()})
table.insert(model, {'conv5_4', nn.SpatialConvolutionMM(512, 512, 3, 3, 1, 1, 1, 1, 1)})
table.insert(model, {'relu5_4', nn.ReLU()})
table.insert(model, {'pool5', inn.SpatialMaxPooling(2, 2, 2, 2)})
table.insert(model, {'torch_view', nn.View(-1):setNumInputDims(3)})
table.insert(model, {'fc6', nn.Linear(25088, 4096)})
table.insert(model, {'relu6', nn.ReLU()})
table.insert(model, {'drop6', nn.Dropout(0.500000)})
table.insert(model, {'fc7', nn.Linear(4096, 4096)})
table.insert(model, {'relu7', nn.ReLU()})
table.insert(model, {'drop7', nn.Dropout(0.500000)})
table.insert(model, {'fc8', nn.Linear(4096, 1000)})
table.insert(model, {'prob', nn.SoftMax()})

**Comment by jcjohnson** [comment]
Hmm, looks like loadcaffe is using `inn` for max pooling even though you requested the `nn` backend. This seems to have been changed in https://github.com/szagoruyko/loadcaffe/commit/7aff9ec16bd80d5db0edce8cab87346928137a08, so maybe you are using an old version of loadcaffe?

Try updating it with `luarocks install loadcaffe`.

**Comment by davidcranor** [comment]
Updated loadcaffe, same problem.

Also, I'm working on setting up an EC2 instance using the install instructions from the neural-style repo (might work better than trying to fix someone else's precompiled system?) and running in CPU mode works.  Got the NVIDIA drivers downloaded and installed, but the nvidia-smi does not see them after reboot.

**Comment by davidcranor** [comment]
Alright, so I followed instructions from

https://github.com/BVLC/caffe/wiki/Install-Caffe-on-EC2-from-scratch-(Ubuntu,-CUDA-7,-cuDNN)

to get CUDA installed on a brand new instance.  I then installed torch, cutorch, cunn, loadcaffe, and neural-style.  I did not install cudnn.  nvidia-smi confirms that the driver is running.

I also had to install `inn` manually.  But it works now!  It's much faster.  Thanks for looking at this.  :-)

I made my EC2 instance public, the ID is ami-23abd746.

**Comment by jcjohnson** [comment]
Glad you figured it out!

**Comment by hughperkins** [comment]
> https://github.com/BVLC/caffe/wiki/Install-Caffe-on-EC2-from-scratch-(Ubuntu,-CUDA-7,-cuDNN)

Thats an excellent page about installing drivers on ec2.  Nice find :-)

**Comment by dmgottlieb** [comment]
@davidcranor I have not been able to find your EC2 image, are you sure it's public? Thanks for doing this.

**Comment by davidcranor** [comment]
@dmgottleib I think it's public....?  Sorry, i am new at this.  Attached is an image of what my dashboard shows....

<img width="1064" alt="screen shot 2015-09-24 at 6 09 23 pm" src="https://cloud.githubusercontent.com/assets/373589/10088266/9a9b1326-62e7-11e5-9a42-68562527e990.png">

**Comment by davidcranor** [comment]
@hughperkins glad I could contribute :-)

**Comment by dmgottlieb** [comment]
@davidcranor Thanks -- I realized AMIs are listed per-region and I was searching the wrong region.
